File: Day98/optimizedSolution.java
import java.utils.*;

class Solution {
    public ArrayList<Double> getMedian(int[] arr) {
        /**
         * This function returns an array of medians of the elements encountered so far
         * 
         * Approach:
         * 1) Use a maxHeap and minHeap that hold almost equal elements
         * 2) All elements of minHeap must be less than maxHeap
         * 3) Return the extra element or average of minHeap and maxHeap top
         * 
         * Time Complexity: O(n*log(n))
         * Space Complexity: O(n)
         **/

        ArrayList<Double> medians = new ArrayList<>();

        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();

        for(int i = 0; i < arr.length; i++) {
            if(i%2 == 0) {
                boolean flag = false;
                if(!minHeap.isEmpty()) {
                    if(arr[i] > minHeap.peek())
                        flag = true;
                }
                
                if(flag) {
                    maxHeap.add(minHeap.remove());
                    minHeap.add(arr[i]);
                } else {
                    maxHeap.add(arr[i]);
                }
                medians.add((double)maxHeap.peek());
            } else {
                if(arr[i] < maxHeap.peek()) {
                    minHeap.add(maxHeap.remove());
                    maxHeap.add(arr[i]);
                } else {
                    minHeap.add(arr[i]);
                }
                medians.add((double)(minHeap.peek() + maxHeap.peek())/2);
            }

        }
        return medians;
    }
}

File: Day101/optimizedSolution.py
class Solution:

    def nextLargerElement(self, arr):
        '''
        Function to find the next greater element for each element of the array.
        
        Approach:
        1) Iterate the array from the end
        2) Keep track of each element in a stack
        3) Remove from the stack if the element is less than current element
        4) Update the element that greater at the current element after
            adding current element to stack
        5) Update the current element to -1 id stack becomes empty
        
        Time Complexity: O(n)
        Space Complexity: O(n)
        '''
        stack = []

        for i in range(len(arr)-1, -1, -1):
            nextG = -1
            
            while len(stack) > 0:
                top = stack[-1]
                if top > arr[i]:
                    nextG = top
                    break
                else:
                    stack.pop()

            stack.append(arr[i])
            arr[i] = nextG
        
        return arr